File: ruby/boggle.rb
#!/usr/bin/ruby

require './dictionary.rb'

module Boggle
  VERSION = '0.1'
  DICE = [
  "iefyeh",
  "eptslu",
  "henips",
  "lecars",
  "avndze",
  "dkunto",
  "snedwo",
  "tgievn",
  "eylugk",
  "fbirxo",
  "bqajmo", # q is actually 'qu'
  "aaciot",
  "camdep",
  "aybilt",
  "hmosar",
  "wirglu"
  ]
  MIN_WORD_LENGTH = 3
end

SIZE = 4


class Board

  
  def initialize(board)
    if board.length != Boggle.const_get(:DICE).length
      raise ValueError
    end
    @board = board
  end
  
  def self.random(seed=nil)
    if seed
      srand(seed)
    end
    board = []
    for die in Boggle.const_get("DICE") do
      board.insert -1, die[rand(die.length)]
    end
    return Board.new(board)
  end
  
  def to_s
    board = []
    for index in (0...@board.length) do
      if index % SIZE == 0 and index != 0
        board.push "\n"
      end
      board.push @board[index]
    end
    board.join ''
  end
  
  def at(x,y)
    if x >= SIZE or y >= SIZE or x < 0 or y < 0
      return nil
    end
    char = @board[x + SIZE*y]
    char == 'q' ? 'qu' : char
  end
  
end


def solve(board)
  
  d = Dictionary.load()
  
  words = []
  
  def check(d,board,words,x,y,checking,checked)
    if not board.at(x,y) or checked[[x,y]]
      return
    end
    checked[[x,y]] = true
    checking.push(board.at(x,y))
    
    if not d.check_prefix(checking.join '')
      checking.pop()
      checked.delete([x,y])
      return
    end
    
    possible_word = checking.join ''
    if possible_word.length >= Boggle.const_get(:MIN_WORD_LENGTH) and d.check_word(possible_word)
      words.push(possible_word)
    end
    
    for y_delta in [-1,0,1] do
      for x_delta in [-1,0,1]
        check(d,board,words,x+x_delta,y+y_delta,checking,checked)
      end
    end  
    checked.delete([x,y])
    checking.pop()
  end
  
  for x in (0...SIZE)
    for y in (0...SIZE)
      check(d,board,words,x,y,[],{})
    end
  end
    
  return words
end
  

def main(args)
  if args.length > 0
    if args[0].length == 16
      STDERR.puts "Using #{args[0]} as board...\n"
      b = Board.new(args[0])
    else
      STDERR.puts "Using random board with seed '#{args[0]}'" +
          " (If you want to specify a board it must be 16 chars.)"
      b = Board.random(args[0])
    end
    
  else
    STDERR.puts "Using random board..."
    b = Board.random()
  end
  
  solved = solve(b)
  solved = solved.to_set().to_a().sort()
  puts solved
end
  
main(ARGV)

File: python2/boggle.py
#!/usr/bin/python
import random
import sys
from time import clock
from dictionary import SimpleDictionary as Dictionary

dice = [
"iefyeh",
"eptslu",
"henips",
"lecars",
"avndze",
"dkunto",
"snedwo",
"tgievn",
"eylugk",
"fbirxo",
"bqajmo", # q is actually 'qu'
"aaciot",
"camdep",
"aybilt",
"hmosar",
"wirglu"
]

min_word_length = 3

class Board(object):
	
	size = 4
	
	def __init__(self,board):
		if len(board) != len(dice):
			raise ValueError
		self.board = board 
	
	@staticmethod
	def random(seed=None):
		if seed:
			random.seed(seed)
		board = []
		for die in random.sample(dice,len(dice)):
			board.append(random.choice(die))
		return Board(board)
	
	def __str__(self):
		board = []
		for index,char in enumerate(self.board):
			if index % self.size == 0 and index != 0:
				board.append('\n')
			board.append(char)
		board.append('\n')
		return ''.join(board)
		
	def at(self,x,y):
		if x >= self.size or y >= self.size or x < 0 or y < 0:
			return None
		char = self.board[x+self.size*y]
		return char if char != 'q' else 'qu' # A 'q' is always a 'qu' in boggle
	
		
def solve(board):
	
	d = Dictionary.load()
	
	words = []
	
	def check(x,y,checking,checked):
		if not board.at(x,y) or checked.get((x,y)):
			return
			
		checked[(x,y)] = True
		checking.append(board.at(x,y))
		
		if not d.checkPrefix(checking):
			checking.pop()
			del checked[(x,y)]
			return

		possible_word = ''.join(checking)
			
		if len(possible_word) >= min_word_length and d.checkWord(possible_word):
			words.append(possible_word)
		
		for y_delta in (-1,0,1):
			for x_delta in (-1,0,1):
				check(x+x_delta,y+y_delta,checking,checked)
				
		del checked[(x,y)]
		checking.pop()
		return 

	for x in xrange(board.size):
		for y in xrange(board.size):
			check(x,y,[],{})
	
	return words
	

def main(args):
	if len(args) > 1:
	
		if len(args[1]) == 16:
			sys.stderr.write("Using '%s' as board.\n" % args[1])
			b = Board(args[1])
		else:
			sys.stderr.write("Using random board with seed '%s'. (If you "
				" want to specify a board it must be 16 chars.)\n" % args[1])
			b = Board.random(args[1])
	else:
		sys.stderr.write("Using random board.\n")
		b = Board.random()
	
	sys.stderr.write(str(b) + "\n")
	solved = list(set(solve(b)))
	solved.sort()
	for s in solved:
		print s
	
if __name__ == "__main__":
	main(sys.argv)
		
			